### PR TITLE
Feature mouse event

### PR DESCRIPTION
**Pull request by StuartSulaski**

### CODE REVIEW

File: include/lug/Window/Event.hpp [hunk 1]
```diff
@@ -12,11 +14,35 @@ namespace Window {
  * @brief      Represents a keyboard event
  */
 struct LUG_WINDOW_API KeyEvent {
-    Keyboard::Key code; ///< The key that triggered the event
-    bool alt;           ///< True if alt was pressed, False otherwise
-    bool ctrl;          ///< True if ctrl was pressed, False otherwise
-    bool shift;         ///< True if shift was pressed, False otherwise
-    bool system;        ///< True if system was pressed, False otherwise
+    Keyboard::Key code;   ///< The key that triggered the event
+    bool alt;                              ///< True if alt was pressed, False otherwise
```
Comment (Quentin01) [inline include/lug/Window/Event.hpp]: There is a problem of indentation for the comments. The same applies for the comments in `MouseEvent`.

File: include/lug/Window/Event.hpp [hunk 2]
```diff
@@ -12,11 +14,35 @@ namespace Window {
...
+    bool ctrl;                             ///< True if ctrl was pressed, False otherwise
+    bool shift;                            ///< True if shift was pressed, False otherwise
+    bool system;                           ///< True if system was pressed, False otherwise
+};
+
+/**
+* @brief      Represents a mouse event
+*/
+struct LUG_WINDOW_API MouseEvent {
+    union {
```
Comment (Quentin01) [inline include/lug/Window/Event.hpp]: Use vertical spaces to between `struct` and `union` according to the guidelines.

File: include/lug/Window/Event.hpp [hunk 3]
```diff
@@ -35,11 +61,17 @@ struct LUG_WINDOW_API Event {
...
+        Close,              ///< Window close event
+        Destroy,            ///< Window destroy event
+        KeyPressed,         ///< KeyPressed event
+        KeyReleased,        ///< KeyReleased event
+        CharEntered,        ///< CharEntered event
+        ButtonPressed,      ///< ButtonPressed event
+        ButtonReleased,     ///< ButtonReleased event
+        MouseMoved,         ///< MouseMoved event
+        MouseWheel,     ///< MouseWheelRotated event
+        MouseLeave,
```
Comment (Quentin01) [inline include/lug/Window/Event.hpp]: Missing comments for `MouseLeave` and `MouseEnter`

File: include/lug/Window/Event.hpp [hunk 4]
```diff
@@ -35,11 +61,17 @@ struct LUG_WINDOW_API Event {
...
-        CharEntered,    ///< CharEntered event
+        Close,              ///< Window close event
+        Destroy,            ///< Window destroy event
+        KeyPressed,         ///< KeyPressed event
+        KeyReleased,        ///< KeyReleased event
+        CharEntered,        ///< CharEntered event
+        ButtonPressed,      ///< ButtonPressed event
+        ButtonReleased,     ///< ButtonReleased event
+        MouseMoved,         ///< MouseMoved event
+        MouseWheel,     ///< MouseWheelRotated event
```
Comment (Quentin01) [inline include/lug/Window/Event.hpp]: There is a problem of indentation for the comment.

File: include/lug/Window/Event.hpp [hunk 5]
```diff
@@ -50,6 +82,7 @@ struct LUG_WINDOW_API Event {
     union {
         KeyEvent key;           ///< A KeyEvent
         CharEvent character;    ///< A CharEvent
+        MouseEvent button;      ///< A MouseEvent
```
Comment (Quentin01) [inline include/lug/Window/Event.hpp]: I don't think `button` is a good name for this one. Maybe `mouse`.

File: include/lug/Window/Event.hpp [hunk 6]
```diff
@@ -12,11 +14,35 @@ namespace Window {
...
+            double yOffset;
+        } scrollOffset;     ///< The distance in x and y that the scroll wheel has traveled
+    };
+    struct {
+        int32_t x;
+        int32_t y;
+    } coord;                ///< The mouse coordinate if the event was triggered by movement
+    bool ctrl;     ///< True if ctrl was pressed, False otherwise
+    bool shift;    ///< True if shift was pressed, False otherwise
+    bool lMouse;   ///< True if left mouse button was pressed, False otherwise
```
Comment (Quentin01) [inline include/lug/Window/Event.hpp]: Why do you need to suffix all the buttoons with `Mouse`? I don't think it's the right solution.

File: include/lug/Window/Mouse.hpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+enum class LUG_WINDOW_API Button : int {
+    Unknown,    // Unhandled button
+
+    Left,       // The left mouse button
+    Right,      // The right mouse button
+    Middle,     // The middle (wheel) mouse button
+    XButton1,   // The first extra mouse button
+    XButton2,   // The second extra mouse button
+
+    ENUM_LENGTH
```
Comment (Quentin01) [inline include/lug/Window/Mouse.hpp]: Why `ENUM_LENGTH`?

File: include/lug/Window/Window.hpp [hunk 8]
```diff
@@ -115,6 +115,24 @@ class LUG_WINDOW_API Window {
...
+    bool isMousePressed(Mouse::Button button) const;
+
+    /**
+    * @brief      Used to retrieve the mouses position
+    *
+    * @param[in]  x   The parametre that will be set to the mouse's position on the x axis.
+    * @param[in]  y   The parametre that will be set to the mouse's position on the x axis.
+    *
+    */
+    void getMousePos(uint32_t & x, uint32_t & y) const;
```
Comment (Quentin01) [inline include/lug/Window/Window.hpp]: We don't use spaces between the type and `&` or `*` in the guidelines.

File: include/lug/Window/Window.hpp [hunk 9]
```diff
@@ -115,6 +115,24 @@ class LUG_WINDOW_API Window {
...
+    * @param[in]  button   The button to check.
+    *
+    * @return     True if the button is pressed, False otherwise.
+    */
+    bool isMousePressed(Mouse::Button button) const;
+
+    /**
+    * @brief      Used to retrieve the mouses position
+    *
+    * @param[in]  x   The parametre that will be set to the mouse's position on the x axis.
```
Comment (Quentin01) [inline include/lug/Window/Window.hpp]: `parameter` maybe?

File: include/lug/Window/Window.hpp [hunk 10]
```diff
@@ -140,6 +163,18 @@ class LUG_WINDOW_API Window {
      */
     std::unordered_map<Keyboard::Key, bool> _keyState;
 
+    /**
+    * Map to store the current state of the mouse buttons, used by #isMousePressed.
+    */
+    std::unordered_map<Mouse::Button, bool> _mouseState;
+
+    /**
+    * Ued to store the mouse's position, used by #getMousePos.
```
Comment (Quentin01) [inline include/lug/Window/Window.hpp]: `Used`

File: samples/window/main.cpp [hunk 11]
```diff
@@ -197,6 +218,15 @@ int main() {
                 window->close();
             }
 
+//            // Checking to see if mouse position is stored correctly
```
Comment (Quentin01) [inline samples/window/main.cpp]: Do we need to keep this source code in comment?

File: samples/window/main.cpp [hunk 12]
```diff
@@ -207,6 +237,51 @@ int main() {
...
+            // Checking ButtonReleased events
+            if (event.type == lug::Window::Event::Type::ButtonReleased) {
+                logger->info(buttonEnumMap[event.button.code] + " is released at position {} {}", event.button.coord.x, event.button.coord.y);
+            }
+            
+            // Checking MouseWheel events
+            if (event.type == lug::Window::Event::Type::MouseWheel) {
+                logger->info("Mouse scroll wheel offset {} {}", event.button.scrollOffset.xOffset, event.button.scrollOffset.yOffset);
+                if (event.button.ctrl) {
+                    logger->info("MODIFIER CTRL");
```
Comment (Quentin01) [inline samples/window/main.cpp]: Do we need this text? If yes, maybe we can change it to something more "esthetic".

File: samples/window/main.cpp [hunk 13]
```diff
@@ -207,6 +237,51 @@ int main() {
...
+            }
+
+            // Checking ButtonReleased events
+            if (event.type == lug::Window::Event::Type::ButtonReleased) {
+                logger->info(buttonEnumMap[event.button.code] + " is released at position {} {}", event.button.coord.x, event.button.coord.y);
+            }
+            
+            // Checking MouseWheel events
+            if (event.type == lug::Window::Event::Type::MouseWheel) {
+                logger->info("Mouse scroll wheel offset {} {}", event.button.scrollOffset.xOffset, event.button.scrollOffset.yOffset);
```
Comment (Quentin01) [inline samples/window/main.cpp]: See the guidelines concerning vertical spaces. You need to indent between `if`.

File: samples/window/main.cpp [hunk 14]
```diff
@@ -207,6 +237,51 @@ int main() {
...
+                logger->info("Mouse just entered the window.");
+                if (event.button.ctrl) {
+                    logger->info("MODIFIER CTRL");
+                }
+                if (event.button.shift) {
+                    logger->info("MODIFIER SHIFT");
+                }
+            }
+
+//            // Checking MouseMoved events
```
Comment (Quentin01) [inline samples/window/main.cpp]: Do we need to keep this source code in comment?

File: samples/window/main.cpp [hunk 15]
```diff
@@ -236,6 +311,13 @@ int main() {
 //                logger->info(it->second + " is set to pressed");
 //            }
 //        }
+
+//        // Checking that mouse button states are correctly set
```
Comment (Quentin01) [inline samples/window/main.cpp]: Do we need to keep this source code in comment?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 16]
```diff
@@ -16,11 +18,11 @@ bool priv::WindowImpl::create(const std::string& title, Style style) {
         return false;
     }
 
-    int minor = XkbMinorVersion;
+    /*int minor = XkbMinorVersion;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Why have you comment that?
Maybe we can just remove it.

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 17]
```diff
@@ -58,6 +66,15 @@ void priv::WindowImpl::close() {
     }
 }
 
+static Mouse::Button buttoncodeToLugButton(unsigned int buttoncode) {
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: `buttonCodeToLugButton` isn't better?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 18]
```diff
@@ -272,17 +297,77 @@ bool priv::WindowImpl::pollEvent(Event& event) {
...
+                case 7:
+                    event.type = Event::Type::MouseWheel;
+                    event.button.scrollOffset.xOffset = -1;
+                    event.button.scrollOffset.yOffset = 0;
+                    break;
+                default:
+                    event.type = Event::Type::ButtonPressed;
+                    event.button.code = buttoncodeToLugButton(xEvent.xbutton.button);
+            }
+            event.button.coord.x = xEvent.xbutton.x;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Use vertical space after `switch`

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 19]
```diff
@@ -272,17 +297,77 @@ bool priv::WindowImpl::pollEvent(Event& event) {
...
+            break;
+        case ButtonRelease:
+            if (xEvent.xbutton.button != Button4 && xEvent.xbutton.button != Button5 &&
+                xEvent.xbutton.button != 6 && xEvent.xbutton.button != 7) {
+                event.type = Event::Type::ButtonReleased;
+                event.button.code = buttoncodeToLugButton(xEvent.xbutton.button);
+                event.button.coord.x = xEvent.xbutton.x;
+                event.button.coord.y = xEvent.xbutton.y;
+            }
+            else {
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Put `else` on the previous line

File: src/lug/Window/Win32/WindowImplWin32.cpp [hunk 20]
```diff
@@ -377,6 +408,93 @@ void priv::WindowImpl::configKeyEvent(KeyEvent& key, WPARAM wParam, LPARAM lPara
...
+        break;
+    }
+}
+
+void priv::WindowImpl::configMouseWheelEvent(MouseEvent & mouse, UINT message, WPARAM wParam)
+{
+    double xOffset = 0.0;
+    double yOffset = 0.0;
+
+//    if (message == WM_MOUSEWHEEL) {
```
Comment (Quentin01) [inline src/lug/Window/Win32/WindowImplWin32.cpp]: Maybe remove this code.

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 21]
```diff
@@ -272,17 +297,77 @@ bool priv::WindowImpl::pollEvent(Event& event) {
...
+            event.key.ctrl      = (xEvent.xkey.state & ControlMask) != 0;
+            event.key.alt       = (xEvent.xkey.state & Mod1Mask) != 0;
+            event.key.shift     = (xEvent.xkey.state & ShiftMask) != 0;
+            event.key.system    = (xEvent.xkey.state & Mod4Mask) != 0;
+            break;
+        case ButtonPress:
+            switch (xEvent.xbutton.button) {
+                case Button4:
+                    event.type = Event::Type::MouseWheel;
+                    event.button.scrollOffset.xOffset = 0;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Can't you have the exact offset of the scroll like in Windows?

File: src/lug/Window/Window.cpp [hunk 22]
```diff
@@ -22,9 +22,9 @@
 
 namespace lug {
 namespace Window {
-
-Window::Window() : _impl{new priv::WindowImpl(this)} {
+    Window::Window() : _impl{new priv::WindowImpl(this)} {
```
Comment (Quentin01) [inline src/lug/Window/Window.cpp]: Why do you have add an indentation?

File: src/lug/Window/Window.cpp [hunk 23]
```diff
@@ -189,6 +189,15 @@ void Window::initKeyState() {
     _keyState[Keyboard::Key::F15] = false;
 }
 
+void Window::initMouseState()
+{
```
Comment (Quentin01) [inline src/lug/Window/Window.cpp]: Bracket at the end of the previous line.

File: src/lug/Window/Window.cpp [hunk 24]
```diff
@@ -208,6 +217,19 @@ bool Window::pollEvent(Event& event) {
         } else if (event.type == Event::Type::KeyReleased) {
             _keyState[event.key.code] = false;
         }
+
+        if (event.type == Event::Type::ButtonPressed) {
+            _mouseState[event.button.code] = true;
+        }
+        else if (event.type == Event::Type::ButtonReleased) {
```
Comment (Quentin01) [inline src/lug/Window/Window.cpp]: `else if` at the end of the previous line.

**Review by Quentin01 — CHANGES_REQUESTED** [review]

File: src/lug/Window/Unix/WindowImplX11.cpp (hunk 21 above)
Comment (ghost) [reply src/lug/Window/Unix/WindowImplX11.cpp]: No, the wheel is like a classic button on X11.

**Review by Quentin01 — APPROVED** [review]

**Pull request merged**